gateway-js/src/loadCsdlFromStorage.ts

```diff
@@ -0,0 +1,102 @@
+import { fetch } from 'apollo-server-env';
+import { GraphQLError } from 'graphql';
+import { CsdlQuery } from './__generated__/graphqlTypes';
+
+export const CSDL_QUERY = /* GraphQL */`#graphql
+  query Csdl($apiKey: String!, $ref: String!) {
+    routerConfig(ref: $ref, apiKey: $apiKey) {
+      __typename
+      ... on RouterConfigResult {
+        id
+        csdl
+      }
+      ... on FetchError {
+        code
+        message
+      }
+    }
+  }
+`;
+
+type CsdlQueryResult = CsdlQuerySuccess | CsdlQueryFailure;
+
+interface CsdlQuerySuccess {
+  data: CsdlQuery;
+}
+
+interface CsdlQueryFailure {
+  data?: CsdlQuery;
+  errors: GraphQLError[];
+}
+
+const { name, version } = require('../package.json');
+
+export async function loadCsdlFromStorage({
+  graphId,
+  graphVariant,
+  apiKey,
+  endpoint,
+  fetcher,
+}: {
+  graphId: string;
+  graphVariant: string;
+  apiKey: string;
+  endpoint: string;
+  fetcher: typeof fetch;
+}) {
+  const result = await fetcher(endpoint, {
+    method: 'POST',
+    body: JSON.stringify({
+      query: CSDL_QUERY,
+      variables: {
+        ref: `${graphId}@${graphVariant}`,
+        apiKey,
+      },
+    }),
+    headers: {
+      'apollographql-client-name': name,
+      'apollographql-client-version': version,
+      'user-agent': `${name}/${version}`,
+      'content-type': 'application/json',
+    },
+  });
+
+
+
+  let response: CsdlQueryResult;
+
+  try {
+    response = await result.json();
+  } catch (e) {
+    // JSON parse error, bad response
```

Comment:
this can also be a network error that happens after headers, right?

Reply:
Can you elaborate? What do you mean by "after headers"?

If there was a network error, I think we'd actually have an unhandled promise rejection in the previous `await fetcher(...)` so I think I should have a try/catch block there as well.

Reply:
Oh we chatted about this on Slack, but: `await fetch()` resolves when the headers have been read, but there can still be network (etc) errors reading the body as part of `await result.json()` (which is why `result.json()` is async and `JSON.parse()` is not).

Reply:
Restructured this a bit for now and going to ask gravity about moving away from our current semantic usage of 400. Let me know if this seems reasonable until then.
9d4f358